src/txmempool.cpp

```diff
@@ -57,46 +57,75 @@ size_t CTxMemPoolEntry::GetTxSize() const
 // Update the given tx for any in-mempool descendants.
 // Assumes that setMemPoolChildren is correct for the given tx and all
 // descendants.
-void CTxMemPool::UpdateForDescendants(txiter updateIt, cacheMap &cachedDescendants, const std::set<uint256> &setExclude)
+void CTxMemPool::UpdateForDescendants(txiter update_it, cacheMap& cache)
 {
-    setEntries stageEntries, setAllDescendants;
-    stageEntries = GetMemPoolChildren(updateIt);
-
-    while (!stageEntries.empty()) {
-        const txiter cit = *stageEntries.begin();
-        setAllDescendants.insert(cit);
-        stageEntries.erase(cit);
-        const setEntries &setChildren = GetMemPoolChildren(cit);
-        for (txiter childEntry : setChildren) {
-            cacheMap::iterator cacheIt = cachedDescendants.find(childEntry);
-            if (cacheIt != cachedDescendants.end()) {
-                // We've already calculated this one, just add the entries for this set
-                // but don't traverse again.
-                for (txiter cacheEntry : cacheIt->second) {
-                    setAllDescendants.insert(cacheEntry);
+    const auto epoch = GetFreshEpoch();
+    // set up the new_cache_line to contain all of our transaction's children (note --
+    // already de-duplicated in case multiple outputs of ours are spent in one
+    // transaction).
+    // Always creates a new cache line entry.
+    vecEntries& new_cache_line = cache[update_it];
+    assert(new_cache_line.empty());
+    // next_idx_to_walk index keeps track of the elements that we've
+    // already expanded.
+    // Invariants:
+    // If index is < next_idx_to_walk, we've already walked it.
+    // If index is >= next_idx_to_walk, we need to walk it.
+    // If next_idx_to_walk >= new_cache_line.size(), we're finished.
+    size_t next_idx_to_walk = 0;
+    txiter next_it = update_it;
+    do {
+        for (const txiter child_it : GetMemPoolChildren(next_it)) {
+            // N.B. children can also be grand children, so guarding
+            // here is neccessary
+            if (visited(child_it)) continue;
+            // if it exists in the cache, copy cached descendants
+            // - All elements in cache are not excluded, no need to check when adding.
+            // - All keys in the cache are already excluded, no need to add
+            //   `child_it` to the new_cache_line (but not all excluded keys
+            //   are in the cache yet!)
+            cacheMap::iterator cached_grand_children = cache.find(child_it);
+            if (cached_grand_children != cache.end()) {
+                for (const txiter grand_child_it : cached_grand_children->second) {
```

Comment:
```suggestion
                for (const txiter& grand_child_it : cached_grand_children->second) {
```

to avoid these compile warnings:
```
txmempool.cpp:89:35: warning: loop variable 'grand_child_it' of type 'const CTxMemPool::txiter' (aka 'const hashed_index_iterator<hashed_index_node<ordered_index_node<boost::multi_index::detail::null_augment_policy, ordered_index_node<boost::multi_index::detail::null_augment_policy, ordered_index_node<boost::multi_index::detail::null_augment_policy, index_node_base<CTxMemPoolEntry, std::allocator<CTxMemPoolEntry> > > > >, boost::multi_index::detail::hashed_unique_tag>, bucket_array<std::allocator<CTxMemPoolEntry> >, boost::multi_index::detail::hashed_index_global_iterator_tag>') creates a copy from type 'const CTxMemPool::txiter' [-Wrange-loop-analysis]
                for (const txiter grand_child_it : cached_grand_children->second) {
                                  ^
txmempool.cpp:89:22: note: use reference type 'const CTxMemPool::txiter &' (aka 'const hashed_index_iterator<hashed_index_node<ordered_index_node<boost::multi_index::detail::null_augment_policy, ordered_index_node<boost::multi_index::detail::null_augment_policy, ordered_index_node<boost::multi_index::detail::null_augment_policy, index_node_base<CTxMemPoolEntry, std::allocator<CTxMemPoolEntry> > > > >, boost::multi_index::detail::hashed_unique_tag>, bucket_array<std::allocator<CTxMemPoolEntry> >, boost::multi_index::detail::hashed_index_global_iterator_tag> &') to prevent copying
                for (const txiter grand_child_it : cached_grand_children->second) {
```